core/rapid_test.go

```diff
@@ -57,183 +57,66 @@ func (m *mockInsertedProposals) insertProposal(
-
-		// Set the multicast callback to relay the message
-		// to the entire cluster
-		multicastFn = func(message *proto.Message) {
-			cluster.pushMessage(message)
-		}
-
-		// Run the sequence up until a certain height
-		for height := uint64(0); height < desiredHeight; height++ {
-			// Start the main run loops
-			cluster.runSequence(height)
-
-			// Wait until the main run loops finish
-			cluster.awaitCompletion()
-		}
-
-		// Make sure that the inserted proposal is valid for each height
-		for _, proposalMap := range insertedProposals.proposals {
-			// Make sure the node has the adequate number of inserted proposals
-			assert.Len(t, proposalMap, int(desiredHeight))
-
-			for _, insertedProposal := range proposalMap {
-				assert.Equal(t, correctRoundMessage.proposal, insertedProposal)
-			}
-		}
-	})
+	// desiredHeight is the desired height number
+	desiredHeight uint64
 }
 
-// getByzantineNodes returns a random subset of
-// byzantine nodes
-func getByzantineNodes(
-	numNodes uint64,
-	set [][]byte,
-) map[string]struct{} {
-	gen := rapid.SampledFrom(set)
-	byzantineNodes := make(map[string]struct{})
-
-	for i := 0; i < int(numNodes); i++ {
-		byzantineNodes[string(gen.Example(i))] = struct{}{}
+// generatePropertyTestEvent generates propertyTestEvent model
+func generatePropertyTestEvent(t *rapid.T) *propertyTestEvent {
+	var (
+		numNodes             = rapid.Uint64Range(4, 15).Draw(t, "number of cluster nodes")
+		numByzantineNodes    = rapid.Uint64Range(0, maxFaulty(numNodes)).Draw(t, "number of byzantine nodes")
+		silentByzantineNodes = rapid.Uint64Range(0, numByzantineNodes).Draw(t, "number of silent byzantine nodes")
+		desiredHeight        = rapid.Uint64Range(10, 20).Draw(t, "minimum height to be reached")
+	)
+
+	return &propertyTestEvent{
+		nodes:                numNodes,
+		byzantineNodes:       numByzantineNodes,
+		silentByzantineNodes: silentByzantineNodes,
+		badByzantineNodes:    numByzantineNodes - silentByzantineNodes,
+		desiredHeight:        desiredHeight,
 	}
-
-	return byzantineNodes
 }
 
-// TestProperty_MajorityHonestNodes is a property-based test
-// that assures the cluster can reach consensus on any
-// arbitrary number of valid nodes and byzantine nodes
-func TestProperty_MajorityHonestNodes(t *testing.T) {
+// TestProperty is a property-based test
+// that assures the cluster can handle rounds properly in any cases.
+func TestProperty(t *testing.T) {
 	t.Parallel()
 
 	rapid.Check(t, func(t *rapid.T) {
 		var multicastFn func(message *proto.Message)
 
 		var (
-			numNodes          = rapid.Uint64Range(4, 30).Draw(t, "number of cluster nodes")
-			numByzantineNodes = rapid.Uint64Range(1, maxFaulty(numNodes)).Draw(t, "number of byzantine nodes")
-			desiredHeight     = rapid.Uint64Range(1, 5).Draw(t, "minimum height to be reached")
-
-			nodes             = generateNodeAddresses(numNodes)
-			insertedProposals = newMockInsertedProposals(numNodes)
+			testEvent         = generatePropertyTestEvent(t)
```

Comment:
do we have only single event per all rounds?

Reply:
Yes. Do you think it makes sense to use different messages? Or do you mean using a different setup for each round?

Reply:
If we'd be able to generate like a sequence of events for each round it'd let us to cover much more test cases. Just check:
1. Select N that stands for number of validators
2. Each of them during a round can behave according to consensus or as a malicious user (don't answer at all, answer with an arbitrary delay, everything malicious we can imagine).
3. For each round, we can count the number of normal and malicious users and check properties accordingly. If we have a consensus, then the round should end successfully. If not, then it should get stuck, I believe.

I believe we can do this as a next step, though.

Reply:
Decided to do it in a new PR.